chainer/optimizer.py

```diff
@@ -27,153 +30,314 @@ def exponential_decay_noise(xp, shape, dtype, hook, opt):
 
-        .. seealso:: :meth:`init_state_cpu`, :meth:`init_state_gpu`
+        Args:
+            param (~chainer.Variable): Variable to be updated.
 
         """
         with cuda.get_device(param.data) as dev:
             if int(dev) == -1:
-                self.init_state_cpu(param, state)
+                self.update_core_cpu(param)
             else:
-                self.init_state_gpu(param, state)
+                self.update_core_gpu(param)
 
-    def init_state_cpu(self, param, state):
-        """Initializes the optimizer state on CPU.
+    def update_core_cpu(self, param):
+        """Updates the parameter on CPU.
 
-        This method is called from :meth:`init_state` by default.
+        See :meth:`update_core` for details.
 
         Args:
-            param (~chainer.Variable): Parameter variable. Its data array is
-                of type :class:`numpy.ndarray`.
-            state (dict): State dictionary.
+            param (~chainer.Variable): Variable to be updated.
+
+        """
+        raise NotImplementedError
+
+    def update_core_gpu(self, param):
+        """Updates the parameter on GPU.
+
+        See :meth:`update_core` for details.
 
-        .. seealso:: :meth:`init_state`
+        Args:
+            param (~chainer.Variable): Variable to be updated.
+
+        """
+        raise NotImplementedError
+
+    def init_state(self, param):
+        """Initializes the state.
+
+        Any implementations that use the state should override this mehtod.
+        This method is called at the first update.
+
+        Args:
+            param (~chainer.Variable): Parameter variable. It can be used to
+                extract the shape and the data type of the parameter.
 
         """
         pass
 
-    def init_state_gpu(self, param, state):
-        """Initializes the optimizer state on GPU.
+    def serialize(self, serializer):
+        """Serializes the update rule state.
 
-        This method is called from :meth:`init_state` by default.
+        Be careful that this method only saves/loads the state of the update
+        rule.
 
         Args:
-            param (~chainer.Variable): Parameter variable. Its data array is
-                of type :class:`cupy.ndarray`.
-            state (dict): State dictionary.
+            serializer (~chainer.AbstractSerializer): Serializer object.
+
+        """
+        if self.state is None:
+            if isinstance(serializer, serializer_module.Deserializer):
+                # try to initialize the state to retrieve state entries
+                self._state = {}
+                self_copy = copy.copy(self)
+                arr = numpy.empty(1, dtype=numpy.float32)
+                self_copy.init_state(chainer.Variable(arr, grad=arr))
+
+                for key in self._state:
+                    self._state[key] = serializer(key, None)
+        else:
+            for key in self._state:
+                self._state[key] = serializer(key, self._state[key])
```

Comment:
We have to assign it here. We cannot assume that `serializer` here is `Serializer`. It can be a deserializer (be careful that there are two if statements nested here).